util/TLSProxy/Message.pm

```diff
@@ -198,14 +198,14 @@ sub get_messages
         print "  [".$record->decrypt_data."]\n";
     } elsif ($record->content_type == TLSProxy::Record::RT_ALERT) {
         my ($alertlev, $alertdesc) = unpack('CC', $record->decrypt_data);
-        #All alerts end the test
-        $end = 1;
         #A CloseNotify from the client indicates we have finished successfully
         #(we assume)
-        if (!$server && $alertlev == AL_LEVEL_WARN
+        if (!$end && !$server && $alertlev == AL_LEVEL_WARN
             && $alertdesc == AL_DESC_CLOSE_NOTIFY) {
             $success = 1;
         }
+        #All alerts end the test
+        $end = 1;
```

Comment:
The client sends two different alerts. I believe it's because the connection isn't considered shut down after a fatal alert already. So if you manage to call two operations that fail (say two "parallel" `SSL_read` and `SSL_write` calls fail) or `SSL_shutdown` after a fatal alert, I believe OpenSSL will send two of them.

See the linked BoringSSL change for what I ended up doing. Messing with `ssl->shutdown` doesn't quite work because that interacts with all the rest of the exposed shutdown state. (`SSL_set_shutdown` is particularly annoying.)